File: backend/app/services/tipsa.py
"""
TIPSA mapping service
"""
from typing import List
from app.models.order import OrderStandard, TIPSAOrder, MappingResponse
from app.core.logging import get_logger

logger = get_logger(__name__)


def map_order_to_tipsa(order: OrderStandard, default_service: str = "ESTANDAR") -> TIPSAOrder:
    """
    Map OrderStandard to TIPSA format
    
    Args:
        order: Standardized order
        default_service: Default service type
        
    Returns:
        TIPSA formatted order
    """
    return TIPSAOrder(
        destinatario=order.recipient_name,
        direccion=order.recipient_address,
        cp=order.recipient_postal_code,
        poblacion=order.recipient_city,
        pais=order.recipient_country,
        contacto=order.recipient_contact or order.recipient_name,
        telefono=order.recipient_phone or "",
        email=order.recipient_email or "",
        referencia=order.reference,
        peso=order.weight_kg,
        servicio=default_service,
        bultos=order.packages or "1",
        observaciones=order.observations or "",
        reembolso=order.cash_on_delivery or "0.00",
        nif_consignatario=order.recipient_tax_id or "",
        codigo_cliente=order.customer_code or "",
        multireferencia=order.multi_reference or ""
    )


def map_orders_to_tipsa(orders: List[OrderStandard], default_service: str = "ESTANDAR") -> List[TIPSAOrder]:
    """
    Map multiple orders to TIPSA format
    
    Args:
        orders: List of standardized orders
        default_service: Default service type
        
    Returns:
        List of TIPSA formatted orders
    """
    return [map_order_to_tipsa(order, default_service) for order in orders]


def generate_tipsa_csv(orders: List[OrderStandard], default_service: str = "ESTANDAR") -> str:
    """
    Generate TIPSA CSV content from OrderStandard (legacy method)
    
    Args:
        orders: List of standardized orders
        default_service: Default service type
        
    Returns:
        CSV content as string
    """
    if not orders:
        return ""
    
    tipsa_orders = map_orders_to_tipsa(orders, default_service)
    
    # Extended CSV headers for new fields
    headers = [
        "destinatario",
        "direccion", 
        "cp",
        "poblacion",
        "pais",
        "contacto",
        "telefono",
        "email",
        "referencia",
        "peso",
        "servicio",
        "bultos",
        "observaciones",
        "reembolso",
        "nif_consignatario",
        "codigo_cliente",
        "multireferencia"
    ]
    
    # Generate CSV content
    csv_lines = [";".join(headers)]
    
    for order in tipsa_orders:
        row = [
            order.destinatario,
            order.direccion,
            order.cp,
            order.poblacion,
            order.pais,
            order.contacto,
            order.telefono,
            order.email,
            order.referencia,
            order.peso,
            order.servicio,
            order.bultos,
            order.observaciones,
            order.reembolso,
            order.nif_consignatario,
            order.codigo_cliente,
            order.multireferencia
        ]
        csv_lines.append(";".join(row))
    
    return "\n".join(csv_lines)


def validate_tipsa_data(order: TIPSAOrder) -> List[str]:
    """
    Validate TIPSA order data
    
    Args:
        order: TIPSA order to validate
        
    Returns:
        List of validation errors
    """
    errors = []
    
    # Required fields validation
    if not order.destinatario.strip():
        errors.append("Destinatario is required")
    
    if not order.direccion.strip():
        errors.append("Dirección is required")
    
    if not order.cp.strip():
        errors.append("Código postal is required")
    
    if not order.poblacion.strip():
        errors.append("Población is required")
    
    if not order.pais.strip():
        errors.append("País is required")
    
    if not order.referencia.strip():
        errors.append("Referencia is required")
    
    # Format validation
    if order.cp and not _is_valid_postal_code(order.cp):
        errors.append("Invalid postal code format")
    
    if order.pais and not _is_valid_country_code(order.pais):
        errors.append("Invalid country code")
    
    if order.email and not _is_valid_email(order.email):
        errors.append("Invalid email format")
    
    if order.telefono and not _is_valid_phone(order.telefono):
        errors.append("Invalid phone format")
    
    if order.peso and not _is_valid_weight(order.peso):
        errors.append("Invalid weight format")
    
    return errors


def process_orders_mapping(orders: List[OrderStandard], format_type: str = "csv", service: str = "ESTANDAR") -> MappingResponse:
    """
    Process orders mapping with error handling
    
    Args:
        orders: List of orders to process
        format_type: Output format (csv, json)
        service: Service type
        
    Returns:
        Mapping response with results
    """
    errors = []
    processed_count = 0
    
    try:
        if format_type == "csv":
            data = generate_tipsa_csv(orders, service)
        elif format_type == "json":
            tipsa_orders = map_orders_to_tipsa(orders, service)
            data = "\n".join([order.model_dump_json() for order in tipsa_orders])
        else:
            errors.append(f"Unsupported format: {format_type}")
            return MappingResponse(
                success=False,
                data="",
                format=format_type,
                count=0,
                errors=errors
            )
        
        processed_count = len(orders)
        
        logger.log_business_event(
            "orders_mapped",
            count=processed_count,
            format=format_type,
            service=service
        )
        
        return MappingResponse(
            success=True,
            data=data,
            format=format_type,
            count=processed_count,
            errors=errors
        )
        
    except Exception as e:
        logger.log_error(e, "process_orders_mapping")
        errors.append(f"Processing error: {str(e)}")
        
        return MappingResponse(
            success=False,
            data="",
            format=format_type,
            count=processed_count,
            errors=errors
        )




def _is_valid_postal_code(cp: str) -> bool:
    """Validate Spanish postal code format"""
    import re
    return bool(re.match(r'^\d{5}$', cp.strip()))


def _is_valid_country_code(country: str) -> bool:
    """Validate ISO country code format"""
    import re
    return bool(re.match(r'^[A-Z]{2}$', country.strip().upper()))


def _is_valid_email(email: str) -> bool:
    """Validate email format"""
    import re
    return bool(re.match(r'^[^\s@]+@[^\s@]+\.[^\s@]+$', email.strip()))


def _is_valid_phone(phone: str) -> bool:
    """Validate Spanish phone format"""
    import re
    cleaned = phone.strip().replace(" ", "")
    return bool(re.match(r'^(\+34|0034)?[6-9]\d{8}$', cleaned))


def _is_valid_weight(weight: str) -> bool:
    """Validate weight format"""
    try:
        weight_num = float(weight)
        return 0 < weight_num <= 1000  # Max 1000kg
    except ValueError:
        return False

File: extension/src/mappers/tipsa.ts
import { OrderStandard, TIPSAOrder } from '@/common/types'

/**
 * TIPSA CSV mapper configuration
 */
export const TIPSA_CONFIG = {
  headers: [
    'destinatario',
    'direccion', 
    'cp',
    'poblacion',
    'pais',
    'contacto',
    'telefono',
    'email',
    'referencia',
    'peso',
    'servicio'
  ],
  defaultService: 'ESTANDAR',
  defaultWeight: '1.0',
  separator: ';'
} as const

/**
 * Map OrderStandard to TIPSA format
 */
export function mapOrderToTIPSA(order: OrderStandard): TIPSAOrder {
  return {
    destinatario: order.shipping.name,
    direccion: formatAddress(order.shipping),
    cp: order.shipping.postcode,
    poblacion: order.shipping.city,
    pais: order.shipping.country,
    contacto: order.buyer.name,
    telefono: order.buyer.phone || '',
    email: order.buyer.email || '',
    referencia: order.orderId,
    peso: calculateWeight(order),
    servicio: TIPSA_CONFIG.defaultService
  }
}

/**
 * Map multiple orders to TIPSA format
 */
export function mapOrdersToTIPSA(orders: OrderStandard[]): TIPSAOrder[] {
  return orders.map(mapOrderToTIPSA)
}

/**
 * Generate TIPSA CSV content
 */
export function generateTIPSACSV(orders: OrderStandard[]): string {
  const tipsaOrders = mapOrdersToTIPSA(orders)
  
  // Generate CSV header
  const header = TIPSA_CONFIG.headers.join(TIPSA_CONFIG.separator)
  
  // Generate CSV rows
  const rows = tipsaOrders.map(order => 
    TIPSA_CONFIG.headers.map(header => order[header as keyof TIPSAOrder]).join(TIPSA_CONFIG.separator)
  )
  
  return [header, ...rows].join('\n')
}

/**
 * Format shipping address for TIPSA
 */
function formatAddress(shipping: OrderStandard['shipping']): string {
  const parts = [shipping.address1]
  
  if (shipping.address2) {
    parts.push(shipping.address2)
  }
  
  return parts.join(', ')
}

/**
 * Calculate package weight (simplified)
 */
function calculateWeight(order: OrderStandard): string {
  // Simple weight calculation based on number of items
  // In a real implementation, this would use actual product weights
  const totalItems = order.items.reduce((sum, item) => sum + item.qty, 0)
  const baseWeight = 0.5 // Base weight per item in kg
  const totalWeight = totalItems * baseWeight
  
  return Math.max(totalWeight, 0.1).toFixed(1) // Minimum 0.1kg
}

/**
 * Validate TIPSA order data
 */
export function validateTIPSAData(order: TIPSAOrder): { isValid: boolean; errors: string[] } {
  const errors: string[] = []
  
  // Required fields validation
  if (!order.destinatario?.trim()) {
    errors.push('Destinatario is required')
  }
  
  if (!order.direccion?.trim()) {
    errors.push('Dirección is required')
  }
  
  if (!order.cp?.trim()) {
    errors.push('Código postal is required')
  }
  
  if (!order.poblacion?.trim()) {
    errors.push('Población is required')
  }
  
  if (!order.pais?.trim()) {
    errors.push('País is required')
  }
  
  if (!order.referencia?.trim()) {
    errors.push('Referencia is required')
  }
  
  // Format validation
  if (order.cp && !isValidPostalCode(order.cp)) {
    errors.push('Invalid postal code format')
  }
  
  if (order.pais && !isValidCountryCode(order.pais)) {
    errors.push('Invalid country code')
  }
  
  if (order.email && !isValidEmail(order.email)) {
    errors.push('Invalid email format')
  }
  
  if (order.telefono && !isValidPhone(order.telefono)) {
    errors.push('Invalid phone format')
  }
  
  if (order.peso && !isValidWeight(order.peso)) {
    errors.push('Invalid weight format')
  }
  
  return {
    isValid: errors.length === 0,
    errors
  }
}

/**
 * Validate postal code format
 */
function isValidPostalCode(cp: string): boolean {
  // Spanish postal code format: 5 digits
  const spanishCPRegex = /^\d{5}$/
  return spanishCPRegex.test(cp.trim())
}

/**
 * Validate country code format
 */
function isValidCountryCode(country: string): boolean {
  // ISO 3166-1 alpha-2 country codes
  const countryCodeRegex = /^[A-Z]{2}$/
  return countryCodeRegex.test(country.trim().toUpperCase())
}

/**
 * Validate email format
 */
function isValidEmail(email: string): boolean {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
  return emailRegex.test(email.trim())
}

/**
 * Validate phone format
 */
function isValidPhone(phone: string): boolean {
  // Spanish phone format: +34XXXXXXXXX or 9 digits
  const phoneRegex = /^(\+34|0034)?[6-9]\d{8}$/
  return phoneRegex.test(phone.trim().replace(/\s/g, ''))
}

/**
 * Validate weight format
 */
function isValidWeight(weight: string): boolean {
  const weightNum = parseFloat(weight)
  return !isNaN(weightNum) && weightNum > 0 && weightNum <= 1000 // Max 1000kg
}

/**
 * Get TIPSA service options
 */
export function getTIPSAServices(): Array<{ value: string; label: string }> {
  return [
    { value: 'ESTANDAR', label: 'Estándar' },
    { value: 'URGENTE', label: 'Urgente' },
    { value: 'EXPRESS', label: 'Express' },
    { value: 'ECONOMICO', label: 'Económico' }
  ]
}

/**
 * Get TIPSA service label
 */
export function getTIPSAServiceLabel(service: string): string {
  const services = getTIPSAServices()
  const found = services.find(s => s.value === service)
  return found?.label || service
}

/**
 * Export TIPSA CSV to file
 */
export function exportTIPSACSV(orders: OrderStandard[], filename?: string): void {
  const csvContent = generateTIPSACSV(orders)
  const defaultFilename = `tipsa_orders_${new Date().toISOString().split('T')[0]}.csv`
  
  // Create and download file
  const blob = new Blob([csvContent], { type: 'text/csv;charset=utf-8;' })
  const link = document.createElement('a')
  
  if (link.download !== undefined) {
    const url = URL.createObjectURL(blob)
    link.setAttribute('href', url)
    link.setAttribute('download', filename || defaultFilename)
    link.style.visibility = 'hidden'
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    URL.revokeObjectURL(url)
  } else {
    throw new Error('Download not supported in this browser')
  }
}
